keras_eval/metrics.py

```diff
@@ -59,59 +56,58 @@ def metrics_top_k(y_probs, y_true, concepts, top_k, round_decimals=7):
-
-        f1_score = round(2 * utils.safe_divide(precision * sensitivity, precision + sensitivity),
-                         round_decimals)
-        average_f1_score.append(f1_score * total_samples_concept)
-
-        fdr = round(utils.safe_divide(fp, tp + fp), round_decimals)
-        average_fdr.append(fdr * total_samples_concept)
-
-        fpr, tpr, _ = roc_curve(y_true, y_probs[:, idx], pos_label=idx)
-        auroc = round(np.mean(tpr), round_decimals)
-        average_auroc.append(auroc * total_samples_concept)
-
-        metrics['individual'].append({
-            'concept': concept, 'metrics': {'TP': int(tp), 'FP': int(fp), 'FN': int(fn),
-                                            'sensitivity': sensitivity, 'precision': precision,
-                                            'f1_score': f1_score, 'FDR': fdr, 'AUROC': auroc}})
-
-    metrics['average']['confusion_matrix'] = confusion_matrix(y_true, top_preds[:, 0], labels=np.arange(0, len(concepts)))
-    metrics['average']['accuracy'] = utils.round_list(average_accuracy_k)
-    metrics['average']['sensitivity'] = utils.round_list([utils.safe_divide(sum(average_sensitivity), total_samples)],
-                                                         round_decimals)
-    metrics['average']['precision'] = utils.round_list([utils.safe_divide(sum(average_precision), total_samples)],
-                                                       round_decimals)
-    metrics['average']['specificity'] = utils.round_list([utils.safe_divide(sum(average_specificity), total_samples)],
-                                                         round_decimals)
-    metrics['average']['f1_score'] = utils.round_list([utils.safe_divide(sum(average_f1_score), total_samples)],
-                                                      round_decimals)
-    metrics['average']['fdr'] = utils.round_list([utils.safe_divide(sum(average_fdr), total_samples)], round_decimals)
-    metrics['average']['auroc'] = utils.round_list([utils.safe_divide(sum(average_auroc), total_samples)],
-                                                   round_decimals)
-
+        if not np.isnan(precision):
+            average_precision.append(precision * total_samples_concept)
+
+        f1_score = round(2 * utils.safe_divide(precision * sensitivity, precision + sensitivity), round_decimals)
+        if not np.isnan(f1_score):
+            average_f1_score.append(f1_score * total_samples_concept)
+
+        metrics_dict = OrderedDict([('sensitivity', sensitivity), ('precision', precision), ('f1_score', f1_score)])
+
+        # Binary classification metrics
+        if len(concepts) == 2:
+            specificity = round(utils.safe_divide(tn, tn + fp), round_decimals)
+            fdr = round(utils.safe_divide(fp, tp + fp), round_decimals)
+            fpr, tpr, _ = roc_curve(y_true, y_probs[:, idx], pos_label=idx)
+            auroc = round(np.mean(tpr), round_decimals)
```

Comment:
The integrate along the TPR axis is wrong by just computing the mean. 

Instead, can you change it to be `auroc = np.trapz(tpr, fpr)`?

See https://github.com/triagemd/keras-eval/pull/51.

Reply:
Absolutely!! 👍 💯